client/src/main/java/io/pravega/client/stream/impl/EventStreamWriterImpl.java

```diff
@@ -144,13 +155,39 @@
         List<ByteBuffer> data = events.stream().map(serializer::serialize).collect(Collectors.toList());
         CompletableFuture<Void> ackFuture = new CompletableFuture<Void>();
         synchronized (writeFlushLock) {
-            synchronized (writeSealLock) {
-                SegmentOutputStream segmentWriter = getSegmentWriter(routingKey);
-                segmentWriter.write(PendingEvent.withHeader(routingKey, data, ackFuture));
+            if (data.stream().mapToInt(m -> m.remaining()).sum() > Serializer.MAX_EVENT_SIZE) {
+                writeLargeEvent(routingKey, data, ackFuture);
+            } else {
+                synchronized (writeSealLock) {
+                    SegmentOutputStream segmentWriter = getSegmentWriter(routingKey);
+                    segmentWriter.write(PendingEvent.withHeader(routingKey, data, ackFuture));
+                }
             }
         }
         return ackFuture;
     }
+    
+    @GuardedBy("writeFlushLock")
+    private void writeLargeEvent(String routingKey, List<ByteBuffer> events, CompletableFuture<Void> ackFuture) {
+        flush();
+        boolean success = false;
+        LargeEventWriter writer = new LargeEventWriter(UUID.randomUUID(), controller, connectionPool);
+        while (!success) {
+            Segment segment = selector.getSegmentForEvent(routingKey);
+            try {
+                writer.writeLargeEvent(segment, events, tokenProvider, config);
```

Comment:
An `UnsupportedOperationException` would be thrown if the server does not support large appends. Perhaps the client should handle this gracefully. 
[Changing `OLDEST_COMPATIBLE_VERSION` field which causes the SSS to close the connection from an old client would not be a great option here.]

Reply:
Is throwing UnsupportedOperationException with an appropriate error message not graceful? I'm not sure what more we can do.